pkg/kvm/kvm_test.go

```diff
@@ -0,0 +1,438 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright (C) 2023 Intel Corporation
+
+// Package kvm automates plugging of SPDK devices to a QEMU instance
+package kvm
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"log"
+	"net"
+	"os"
+	"path/filepath"
+	"strings"
+	"sync/atomic"
+	"testing"
+	"time"
+
+	pc "github.com/opiproject/opi-api/common/v1/gen/go"
+	pb "github.com/opiproject/opi-api/storage/v1alpha1/gen/go"
+	"github.com/opiproject/opi-spdk-bridge/pkg/frontend"
+	"github.com/opiproject/opi-spdk-bridge/pkg/models"
+	"github.com/opiproject/opi-spdk-bridge/pkg/server"
+	"github.com/ulule/deepcopier"
+	"google.golang.org/protobuf/proto"
+)
+
+var (
+	alwaysSuccessfulJSONRPC    = stubJSONRRPC{nil}
+	alwaysFailingJSONRPC       = stubJSONRRPC{errors.New("stub error")}
+	testVirtioBlkID            = "virtio-blk-42"
+	testCreateVirtioBlkRequest = &pb.CreateVirtioBlkRequest{VirtioBlk: &pb.VirtioBlk{
+		Id:       &pc.ObjectKey{Value: testVirtioBlkID},
+		PcieId:   &pb.PciEndpoint{PhysicalFunction: 42},
+		VolumeId: &pc.ObjectKey{Value: "Malloc42"},
+		MaxIoQps: 1,
+	}}
+	testDeleteVirtioBlkRequest = &pb.DeleteVirtioBlkRequest{Name: testVirtioBlkID}
+	genericQmpError            = `{"error": {"class": "GenericError", "desc": "some error"}}` + "\n"
+	genericQmpOk               = `{"return": {}}` + "\n"
+
+	qmpServerOperationTimeout = 10 * time.Millisecond
+	qmplibTimeout             = 5 * time.Millisecond
+)
+
+type stubJSONRRPC struct {
+	err error
+}
+
+func (s stubJSONRRPC) Call(method string, args, result interface{}) error {
+	if method == "vhost_create_blk_controller" {
+		if s.err == nil {
+			resultCreateVirtioBLk, ok := result.(*models.VhostCreateBlkControllerResult)
+			if !ok {
+				log.Panicf("Unexpected type for virtio-blk device creation result")
+			}
+			*resultCreateVirtioBLk = models.VhostCreateBlkControllerResult(true)
+		}
+		return s.err
+	} else if method == "vhost_delete_controller" {
+		if s.err == nil {
+			resultDeleteVirtioBLk, ok := result.(*models.VhostDeleteControllerResult)
+			if !ok {
+				log.Panicf("Unexpected type for virtio-blk device deletion result")
+			}
+			*resultDeleteVirtioBLk = models.VhostDeleteControllerResult(true)
+		}
+		return s.err
+	} else {
+		return s.err
+	}
+}
+
+type mockCall struct {
```

Comment:
have you considered using https://github.com/vektra/mockery instead ?

Reply:
As I know, all go mock frameworks require code generation and I'd avoid it. In addition as I see other tests are leveraging a goroutine listening on unix socket and sending stub responses and wanted to stick to that approach exercising the full path.

Reply:
ok, just checking